### PR TITLE
Add shared key

### PR DESCRIPTION
**Pull request by paweljakubas**
adp-866

This PR add and refactor several things:
- [x] Add Shared Style which uses purpose=1854H and role=0 and role=2
- [x] Refactor unit test due to change in point 1
- [x] introduce new HRP according to Matthias changes in CIP
- [x] Commands Inspect/Hash/Public/Child are aligned 
- [x] Shared style is introduced on command level (FromRecoveryPhase)
- [x] Command tests are aligned 
- [x] Update README.md

### CODE REVIEW

File: command-line/lib/Command/Key/Child.hs [hunk 1]
```diff
@@ -116,39 +119,55 @@ run Child{path} = do
...
     -- as m / purpose' or m / purpose' / coin_type' so we do not expose them.
     childHrpFor [_,_,_,2,_] hrp
         | hrp == CIP5.root_xsk = pure CIP5.stake_xsk
-
-    childHrpFor [_,_,_,3,_] hrp
-        | hrp == CIP5.root_xsk = pure CIP5.script_xsk
-
-    childHrpFor [_,_,_,4,_] hrp
-        | hrp == CIP5.root_xsk = pure CIP5.script_xsk
+        | hrp == CIP5.shared_root_xsk = pure CIP5.shared_stake_xsk
```
Comment (KtorZ) [inline command-line/lib/Command/Key/Child.hs:149]: :+1:

File: command-line/lib/Command/Key/Child.hs [hunk 2]
```diff
@@ -116,39 +119,55 @@ run Child{path} = do
...
 
     childHrpFor [_,_,_,_,_] hrp
         | hrp == CIP5.root_xsk = pure CIP5.addr_xsk
+        | hrp == CIP5.shared_root_xsk = pure CIP5.shared_addr_xsk
```
Comment (KtorZ) [inline command-line/lib/Command/Key/Child.hs:153]: :+1:

File: command-line/lib/Command/Key/Child.hs [hunk 3]
```diff
@@ -116,39 +119,55 @@ run Child{path} = do
...
 
     childHrpFor [_,_,_] hrp
         | hrp == CIP5.root_xsk = pure CIP5.acct_xsk
+        | hrp == CIP5.shared_root_xsk = pure CIP5.shared_acct_xsk
```
Comment (KtorZ) [inline command-line/lib/Command/Key/Child.hs:157]: :+1:

File: command-line/lib/Command/Key/Child.hs [hunk 4]
```diff
@@ -116,39 +119,55 @@ run Child{path} = do
...
-
-    childHrpFor [3,_] hrp
-        | hrp == CIP5.acct_xsk = pure CIP5.script_xsk
-        | hrp == CIP5.acct_xvk = pure CIP5.script_xvk
-
-    childHrpFor [4,_] hrp
-        | hrp == CIP5.acct_xsk = pure CIP5.script_xsk
-        | hrp == CIP5.acct_xvk = pure CIP5.script_xvk
+        | hrp == CIP5.shared_acct_xsk = pure CIP5.shared_stake_xsk
+        | hrp == CIP5.shared_acct_xvk = pure CIP5.shared_stake_xvk
```
Comment (KtorZ) [inline command-line/lib/Command/Key/Child.hs:163]: :+1:

File: command-line/lib/Command/Key/Child.hs [hunk 5]
```diff
@@ -116,39 +119,55 @@ run Child{path} = do
...
 
     childHrpFor [_,_] hrp
         | hrp == CIP5.root_xsk = pure CIP5.addr_xsk
         | hrp == CIP5.acct_xsk = pure CIP5.addr_xsk
         | hrp == CIP5.acct_xvk = pure CIP5.addr_xvk
+        | hrp == CIP5.shared_acct_xsk = pure CIP5.shared_addr_xsk
+        | hrp == CIP5.shared_acct_xvk = pure CIP5.shared_addr_xvk
```
Comment (KtorZ) [inline command-line/lib/Command/Key/Child.hs:170]: :+1:

File: command-line/lib/Command/Key/Child.hs [hunk 6]
```diff
@@ -158,6 +177,12 @@ run Child{path} = do
             \(purpose, coin_type, account, role, index) if you're dealing with \
             \anything else."
 
+        | hrp == CIP5.shared_root_xsk = fail
+            "When deriving child keys from a parent root key, you must \
+            \provide either 3 or 5 path segments. Provide 3 \
+            \(purpose, coin_type, account) or 5 \
+            \(purpose, coin_type, account, role, index)."
```
Comment (KtorZ) [inline command-line/lib/Command/Key/Child.hs:184]: Indeed.

File: command-line/lib/Command/Key/Hash.hs [hunk 7]
```diff
@@ -68,13 +70,16 @@ run Hash = do
...
+        | hrp == CIP5.stake_vk         = CIP5.stake_vkh
+        | hrp == CIP5.stake_xvk        = CIP5.stake_vkh
 
-        | hrp == CIP5.script_vk  = CIP5.script_vkh
-        | hrp == CIP5.script_xvk = CIP5.script_vkh
+        | hrp == CIP5.shared_addr_vk   = CIP5.script_vkh
+        | hrp == CIP5.shared_addr_xvk  = CIP5.script_vkh
+
+        | hrp == CIP5.shared_stake_vk  = CIP5.script_vkh
+        | hrp == CIP5.shared_stake_xvk = CIP5.script_vkh
```
Comment (KtorZ) [inline command-line/lib/Command/Key/Hash.hs:83]: I'd be in favor of dropping the `script_` prefix to use something more consistent like `addr_shared_vkh` and `stake_shared_vkh`

File: core/lib/Cardano/Address/Style/Shared.hs [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+--
+-- @since 3.4.0
+deriveMultisigForDelegationPublicKey
+    :: Shared 'AccountK XPub
+    -> Index 'Soft 'PaymentK
+    -> Shared 'ScriptK XPub
+deriveMultisigForDelegationPublicKey accPub addrIx =
+    let (Shared xpub) = Internal.deriveAddressPublicKey accPub Stake addrIx
+    in Shared xpub
+
```
Comment (KtorZ) [inline core/lib/Cardano/Address/Style/Shared.hs:280]: We could stick with:

```
    , deriveAccountPrivateKey
    , deriveAddressPrivateKey
    , deriveDelegationPrivateKey
    , deriveAddressPublicKey
```

For all these to be consistent with Shelle and Icarus wallets. In the end, the "multisig" aspect is implied by the Shared key already.

File: core/lib/Cardano/Address/Style/Shelley.hs [hunk 9]
```diff
@@ -39,10 +39,6 @@ module Cardano.Address.Style.Shelley
     , deriveAddressPrivateKey
     , deriveDelegationPrivateKey
     , deriveAddressPublicKey
-    , deriveMultisigForPaymentPrivateKey
-    , deriveMultisigForPaymentPublicKey
-    , deriveMultisigForDelegationPrivateKey
-    , deriveMultisigForDelegationPublicKey
```
Comment (KtorZ) [inline core/lib/Cardano/Address/Style/Shelley.hs:45]: :+1:

File: core/test/Cardano/Address/ScriptSpec.hs [hunk 10]
```diff
@@ -85,31 +85,27 @@ import qualified Data.Text.Encoding as T
 
 spec :: Spec
 spec = do
-    let mnemonic = [ "test", "child", "burst", "immense", "armed", "parrot"
-                   , "company", "walk", "dog" ]
-    let (Right mw) = mkSomeMnemonic @'[9,12,15,18,21,24] mnemonic
+    let mnemonic = ["network","empty","cause","mean","expire","private","finger"
+                   ,"accident","session","problem","absurd","banner","stage","void","what"]
+    let (Right mw) = mkSomeMnemonic @'[15] mnemonic
```
Comment (KtorZ) [inline core/test/Cardano/Address/ScriptSpec.hs:90]: Why this change :thinking:  ?

File: core/lib/Cardano/Address/Style/Shared.hs (hunk 8 above)
Comment (paweljakubas) [reply core/lib/Cardano/Address/Style/Shared.hs:280]: makes sense. will do it

File: command-line/lib/Command/Key/Hash.hs (hunk 7 above)
Comment (rvl) [reply command-line/lib/Command/Key/Hash.hs]: So, we would also rename `shared_addr_vk` to `addr_shared_vk` to match, right?

I pushed a suggestion to use a mapping of prefixes instead of conditionals.
It makes the code shorter.
Same could be applied in the other CLI commands.

File: core/test/Cardano/Address/ScriptSpec.hs (hunk 10 above)
Comment (paweljakubas) [reply core/test/Cardano/Address/ScriptSpec.hs:90]: oh, was double checking in ghci and used 15-word mnemonic rather than 9-word

**Comment by paweljakubas** [comment]
CIP to comply with https://github.com/cardano-foundation/CIPs/pull/69/files

File: core/lib/Cardano/Address/Style/Shared.hs [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+    ( Word32 )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Address.Derivation as Internal
+import qualified Data.ByteArray as BA
+
+-- $overview
+--
+-- This module provides an implementation of:
```
Comment (rvl) [inline core/lib/Cardano/Address/Style/Shared.hs:82]: Does this module doc need updating?

File: core/lib/Cardano/Address/Style/Shared.hs [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+-- - 'paymentAddress': for constructing payment addresses from a address public key or a script
+-- - 'delegationAddress': for constructing delegation addresses from payment credential (public key or script) and stake credential (public key or script)
+-- - 'pointerAddress': for constructing delegation addresses from payment credential (public key or script) and chain pointer
+-- - 'stakeAddress': for constructing reward accounts from stake credential (public key or script)
+
+-- | A cryptographic key for sequential-scheme address derivation, with
+-- phantom-types to disambiguate key types.
+--
+-- @
+-- let rootPrivateKey = Shelley 'RootK XPrv
```
Comment (rvl) [inline core/lib/Cardano/Address/Style/Shared.hs:97]: This does need updating.

**Review by rvl — COMMENTED** [review]
:+1:

File: command-line/lib/Command/Key/Hash.hs (hunk 7 above)
Comment (KtorZ) [reply command-line/lib/Command/Key/Hash.hs]: > So, we would also rename shared_addr_vk to addr_shared_vk to match, right?

Yes. Sorry, we originally hinted towards `shared_thing` yesterday, but I found the `thing_shared` order a bit more helpful / readable. So, there will be some massive `sed` to do in the end.

File: core/test/Cardano/Address/ScriptSpec.hs (hunk 10 above)
Comment (paweljakubas) [reply core/test/Cardano/Address/ScriptSpec.hs:90]: ok, will return to original mnemonics in coming commit

File: command-line/lib/Command/Key/Hash.hs (hunk 7 above)
Comment (paweljakubas) [reply command-line/lib/Command/Key/Hash.hs]: will do it in the coming commit

File: core/lib/Cardano/Address/Style/Shared.hs [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+-- - 'Cardano.Address.Derivation.GenMasterKey': for generating Shared master keys from mnemonic sentences
+-- - 'Cardano.Address.Derivation.HardDerivation': for hierarchical hard derivation of parent to child keys
+-- - 'Cardano.Address.Derivation.SoftDerivation': for hierarchical soft derivation of parent to child keys
+--
+-- - 'paymentAddress': for constructing payment addresses from a address public key or a script
+-- - 'delegationAddress': for constructing delegation addresses from payment credential (public key or script) and stake credential (public key or script)
+-- - 'pointerAddress': for constructing delegation addresses from payment credential (public key or script) and chain pointer
+-- - 'stakeAddress': for constructing reward accounts from stake credential (public key or script)
+
+-- | A cryptographic key for sequential-scheme address derivation, with
```
Comment (rvl) [inline core/lib/Cardano/Address/Style/Shared.hs:93]: It would be informative to describe this in terms of the shelley scheme. E.g. that it's mostly like Shelley, _except_ ....

File: core/lib/Cardano/Address/Script.hs [hunk 14]
```diff
@@ -209,32 +210,47 @@ scriptHashFromBytes bytes
...
+instance NFData KeyType
+
 -- | A 'KeyHash' type represents verification key hash that participate in building
 -- multi-signature script. The hash is expected to have size of 28-byte.
 --
 -- @since 3.0.0
-newtype KeyHash = KeyHash { unKeyHash :: ByteString }
+data KeyHash = KeyHash
+    { payload :: ByteString
+    , credential :: KeyType }
```
Comment (rvl) [inline core/lib/Cardano/Address/Script.hs:223]: Seems nicer to me to swap the order of records here.

File: command-line/lib/Command/Key/Public.hs [hunk 15]
```diff
@@ -61,22 +61,31 @@ run Public{chainCode} = do
...
         , CIP5.addr_xsk
         , CIP5.stake_xsk
-        , CIP5.script_xsk
+        , CIP5.root_shared_xsk
+        , CIP5.acct_shared_xsk
+        , CIP5.addr_shared_xsk
+        , CIP5.stake_shared_xsk
         ]
 
     prefixToPublic WithChainCode hrp
```
Comment (rvl) [inline command-line/lib/Command/Key/Public.hs:70]: Could you make these into a map lookup, like I did for Hash?

**Review by rvl — COMMENTED** [review]
Looks broadly good. I pushed some cosmetic fixes.
One thing that doesn't quite sit right with me is that the Shared and Shelley address style are very similar, yet they have two separate copies of the derivation code.

**Comment by rvl** [comment]
I rebased the branch on latest master to get some build fixes.

File: core/lib/Cardano/Address/Style/Shared.hs [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+    genMasterKeyFromXPrv = liftXPrv
+    genMasterKeyFromMnemonic fstFactor sndFactor =
+        Shared $ generateNew seedValidated sndFactor
+        where
+            seed  = someMnemonicToBytes fstFactor
+            seedValidated = assert
+                (BA.length seed >= minSeedLengthBytes && BA.length seed <= 255)
+                seed
+
+instance Internal.HardDerivation Shared where
```
[the remark this answers is not in the document]
Comment (KtorZ) [reply core/lib/Cardano/Address/Style/Shared.hs:155]: This would make sense indeed, especially because the rationale behind CIP-1854' really is to stick to the same structure ass CIP-1852'. So if we add some new role to CIP-1852' (e.g. Catalyst Voting Key) we want to also have it in CIP-1854'. Therefore, we do want to share the underlying code as much as possible.

File: command-line/lib/Command/Key/Hash.hs [hunk 17]
```diff
@@ -49,32 +51,25 @@ run Hash = do
...
+    -- Mapping of input HRP to output HRP
+    prefixes =
+        [ ( CIP5.addr_vk         , CIP5.addr_vkh   )
+        , ( CIP5.addr_xvk        , CIP5.addr_vkh   )
+        , ( CIP5.stake_vk        , CIP5.stake_vkh  )
+        , ( CIP5.stake_xvk       , CIP5.stake_vkh  )
+        , ( CIP5.addr_shared_vk  , CIP5.addr_shared_vkh )
+        , ( CIP5.addr_shared_xvk , CIP5.addr_shared_vkh )
+        , ( CIP5.stake_shared_vk , CIP5.stake_shared_vkh )
+        , ( CIP5.stake_shared_xvk, CIP5.stake_shared_vkh )
```
Comment (KtorZ) [inline command-line/lib/Command/Key/Hash.hs:63]: :+1:

File: core/lib/Cardano/Address/Script.hs [hunk 18]
```diff
@@ -209,51 +212,67 @@ scriptHashFromBytes bytes
...
-        | hrp == CIP5.script_vk  = Right $ hashCredential bytes
-        | hrp == CIP5.script_xvk = Right $ hashCredential $ BS.take 32 bytes
-        | otherwise = Left ErrKeyHashFromTextWrongHrp
+        | hrp == CIP5.addr_shared_vkh  = Just (Payment, bytes)
+        | hrp == CIP5.stake_shared_vkh = Just (Delegation, bytes)
+        | hrp == CIP5.addr_shared_vk   = Just (Payment, hashCredential bytes)
+        | hrp == CIP5.addr_shared_xvk  = Just (Payment, hashCredential $ BS.take 32 bytes)
+        | hrp == CIP5.stake_shared_vk  = Just (Delegation, hashCredential bytes)
+        | hrp == CIP5.stake_shared_xvk = Just (Delegation, hashCredential $ BS.take 32 bytes)
+        | otherwise                    = Nothing
```
Comment (KtorZ) [inline core/lib/Cardano/Address/Script.hs:275]: :+1:

File: core/lib/Cardano/Address/Script.hs [hunk 19]
```diff
@@ -319,15 +338,18 @@ validateScript
...
     -> Either ErrValidateScript ()
 validateScript level script = do
-    let validateKeyHash (KeyHash bytes) =
+    let validateKeyHash (KeyHash bytes _) =
             (BS.length bytes == credentialHashSize)
     let allSigs = foldScript (:) [] script
     unless (L.all validateKeyHash allSigs) $ Left WrongKeyHash
 
+    when (L.length (L.nub $ map credential allSigs) > 1) $
+        Left NotUniformKeyType
```
Comment (KtorZ) [inline core/lib/Cardano/Address/Script.hs:347]: :+1:

File: core/lib/Cardano/Address/Script.hs [hunk 20]
```diff
@@ -490,6 +513,8 @@ prettyErrValidateScript = \case
     WrongKeyHash ->
         "The hash of verification key is expected to have "
         <> show credentialHashSize <> " bytes."
+    NotUniformKeyType ->
+        "All key hashes should have either 'addr_shared_vkh' or 'stake_shared_vhk' prefixes."
```
Comment (KtorZ) [inline core/lib/Cardano/Address/Script.hs:517]: Small note: I'd avoid mentioning prefixes here since we also except their `_vk` and `_xvk` variants. Plus, if we change them again, we'll have to change the error message :no_mouth:. Rather, I would say:

> _"All keys of a script must have the same role: either payment or delegation."_

File: core/lib/Cardano/Address/Style/Shelley.hs [hunk 21]
```diff
@@ -962,8 +898,8 @@ unsafeFromRight =
 --- | Computes a 28-byte Blake2b224 digest of a Shelley 'XPub'.
 ---
 --- @since 3.0.0
-hashKey :: Shelley key XPub -> KeyHash
-hashKey = KeyHash . hashCredential . xpubPublicKey . getKey
+hashKey :: KeyType -> Shelley key XPub -> KeyHash
+hashKey cred = flip KeyHash cred . hashCredential . xpubPublicKey . getKey
```
Comment (KtorZ) [inline core/lib/Cardano/Address/Style/Shelley.hs:902]: Does this actually make sense for Shelley keys? The `KeyHash` data-type seems very much tight to monetary scripts (and when serialized, will assume `_shared` prefixes which is probably not what we want. So either the `KeyHash` should be a little more generic (and also, moved out of the `Script` module), or, we should use a different data-type for this. 

We could likely use a phantom type to capture the purpose on the KeyHash. Or simply a data-type proper at the value-level.

File: core/lib/Cardano/Address/Style/Shared.hs [hunk 22]
```diff
@@ -0,0 +1,323 @@
...
+--
+-- > let Just accIx = indexFromWord32 0x80000000
+-- > let acctK = deriveAccountPrivateKey rootK accIx
+-- >
+-- > let Just addIx = indexFromWord32 0x00000014
+-- > let addrK = deriveAddressPrivateKey acctK UTxOExternal addIx
+--
+-- > let stakeK = deriveDelegationPrivateKey acctK
+
+instance Internal.GenMasterKey Shared where
```
[the remark this answers is not in the document]
Comment (paweljakubas) [reply core/lib/Cardano/Address/Style/Shared.hs:143]: reused Shelley and Shared code when it comes to generation and derivation

File: core/lib/Cardano/Address/Style/Shared.hs (hunk 16 above)
[the remark this answers is not in the document]
Comment (paweljakubas) [reply core/lib/Cardano/Address/Style/Shared.hs:155]: reused Shelley and Shared code when it comes to generation and derivation

File: core/lib/Cardano/Address/Style/Shelley.hs (hunk 21 above)
Comment (paweljakubas) [reply core/lib/Cardano/Address/Style/Shelley.hs:902]: it is not needed in Shelley, only in Shared

**Review by KtorZ — APPROVED** [review]
LGTM
